pageserver/src/tenant/timeline.rs

```diff
@@ -818,6 +818,94 @@ impl Timeline {
     pub fn subscribe_for_state_updates(&self) -> watch::Receiver<TimelineState> {
         self.state.subscribe()
     }
+
+    pub fn layer_map_info(&self) -> LayerMapInfo {
+        let mut in_memory_layers = BTreeSet::new();
+        let layer_map = self.layers.read().unwrap();
+        if let Some(open_layer) = &layer_map.open_layer {
+            in_memory_layers.insert(open_layer.info());
+        }
+        for frozen_layer in &layer_map.frozen_layers {
+            in_memory_layers.insert(frozen_layer.info());
+        }
+
+        let mut historic_layers = BTreeSet::new();
+        for historic_layer in layer_map.iter_historic_layers() {
+            historic_layers.insert(historic_layer.info());
+        }
+
+        LayerMapInfo {
+            in_memory_layers,
+            historic_layers,
+        }
+    }
+
+    pub async fn download_layer(&self, layer_file_name: &str) -> anyhow::Result<bool> {
+        if self.remote_client.is_none() {
+            return Ok(false);
+        }
+        let Some(remote_layer) = self.find_layer(layer_file_name)
+            .and_then(PersistentLayer::downcast_remote_layer)
+        else {
+            return Ok(false);
+        };
+
+        self.download_remote_layer(remote_layer)
+            .await
+            .map(|()| true)
+    }
+
+    pub async fn evict_layer(&self, layer_file_name: &str) -> anyhow::Result<bool> {
+        let Some(remote_client) = &self.remote_client
+        else {
+            return Ok(false);
+        };
+        let Some(local_layer) = self.find_layer(layer_file_name)
+        else {
+            return Ok(false);
+        };
+        if local_layer.is_remote_layer() {
+            return Ok(false);
+        }
+
+        // ensure the current layer is uploaded for sure
+        remote_client
+            .wait_completion()
+            .await
+            .context("wait for layer upload ops to complete")?;
+
+        let layer_metadata = LayerFileMetadata::new(
+            local_layer
+                .file_size()
+                .expect("Local layer should have a file size"),
+        );
+        let new_remote_layer = Arc::new(match local_layer.filename() {
+            LayerFileName::Image(image_name) => RemoteLayer::new_img(
+                self.tenant_id,
+                self.timeline_id,
+                &image_name,
+                &layer_metadata,
+            ),
+            LayerFileName::Delta(delta_name) => RemoteLayer::new_delta(
+                self.tenant_id,
+                self.timeline_id,
+                &delta_name,
+                &layer_metadata,
+            ),
+            #[cfg(test)]
+            LayerFileName::Test(_) => unreachable!(),
+        });
+
+        let mut layers = self.layers.write().unwrap();
+        let mut updates = layers.batch_update();
+        local_layer.delete()?;
+        updates.remove_historic(local_layer);
```

Comment:
Checked the impl again.

For level 0 deltas we still have an `Arc::ptr_eq` check which I suspect might be troublesome (but could not prove) -- that'd be easily fixed via `Arc::as_ptr(&left) as *const () == Arc::as_ptr(&right) as *const ()` which the std lib might eventually use as well (no link right now).

This would be one risky place of poisoning the lock, requiring `/ignore` + `/load`, but I think we cannot test this out.

Also, looking at the `insert_historic` this makes me wonder if we could fair with just `insert_historic` + `flush`, but perhaps that'd be bit of a misuse.